File: algorithms/nim/sorting/quick_sort.nim
proc quickSort[T](a: var openArray[T]) =
  if a.len <= 1: return
  
  let pivot = a[a.high]
  var i = 0
  for j in 0 ..< a.high:
    if a[j] < pivot:
      swap(a[i], a[j])
      inc i
  
  swap(a[i], a[a.high])
  
  if i > 0: quickSort(a.toOpenArray(0, i - 1))
  if i < a.high: quickSort(a.toOpenArray(i + 1, a.high))

var nums = @[10, 7, 8, 9, 1, 5]
echo "Original: ", nums
quickSort(nums)
echo "Sorted: ", nums

File: algorithms/java/sorting/QuickSort.java
package algorithms.java.sorting;

import java.util.Arrays;

public class QuickSort {
    
    // Partition method
    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);
        
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                
                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        
        // Swap arr[i+1] and arr[high] (or pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        
        return i + 1;
    }
    
    // Main sort method
    public static void sort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            
            sort(arr, low, pi - 1);
            sort(arr, pi + 1, high);
        }
    }
    
    public static void main(String[] args) {
        int[] arr = {10, 7, 8, 9, 1, 5};
        System.out.println("Original array: " + Arrays.toString(arr));
        
        sort(arr, 0, arr.length - 1);
        
        System.out.println("Sorted array: " + Arrays.toString(arr));
    }
}
